faunadb-java-dsl/src/main/java/com/faunadb/client/types/Properties.java

```diff
@@ -0,0 +1,203 @@
+        return getProperties(clazz, false);
+    }
+
+    static Property[] getWriteProperties(Class<?> clazz) {
+        return getProperties(clazz, true);
+    }
+
+    private static Property[] getProperties(Class<?> clazz, boolean isWrite) {
+        try {
+            Map<String, Field> fields = newHashMap();
+
+            for (Field field : clazz.getDeclaredFields()) {
+                if ((field.getModifiers() & Modifier.STATIC) != 0)
+                    continue;
+
+                if (isWrite && (field.getModifiers() & Modifier.FINAL) != 0)
+                    continue;
+
+                if (field.isAnnotationPresent(FaunaIgnore.class))
+                    continue;
+
+                FaunaField faunaField = field.getAnnotation(FaunaField.class);
+
+                if (faunaField != null) {
+                    if (faunaField.value().length() > 0) {
+                        fields.put(faunaField.value(), field);
+                    } else {
+                        fields.put(field.getName(), field);
+                    }
+                }
+            }
+
+            Map<String, Method> props = newHashMap();
+
+            for (Method method : clazz.getMethods()) {
+                if ((method.getModifiers() & (Modifier.PUBLIC & ~Modifier.STATIC)) == 0)
+                    continue;
+
+                int offset = getPrefix(method, isWrite);
+                if (offset < 0)
+                    continue;
+
+                String propertyName = removePrefix(method.getName(), offset);
+
+                if ("class".equals(propertyName) || method.isAnnotationPresent(FaunaIgnore.class))
+                    continue;
+
+                FaunaField faunaField = method.getAnnotation(FaunaField.class);
+
+                if (faunaField != null && faunaField.value().length() > 0)
+                    propertyName = faunaField.value();
+
+                if (!fields.containsKey(propertyName)) {
+                    props.put(propertyName, method);
+                } else if (faunaField != null) {
+                    fields.remove(propertyName);
+                    props.put(propertyName, method);
+                }
+            }
+
+            List<Property> allProps = newArrayList();
+
+            for (Map.Entry<String, Field> entry : fields.entrySet())
+                allProps.add(new FieldProperty(entry.getKey(), entry.getValue()));
+
+            for (Map.Entry<String, Method> entry : props.entrySet()) {
+                Method method = entry.getValue();
+
+                Type propertyType = isWrite ? method.getGenericParameterTypes()[0] : method.getGenericReturnType();
+                allProps.add(new MethodProperty(entry.getKey(), method, propertyType));
+            }
+
+            return allProps.toArray(new Property[allProps.size()]);
+        } catch (FaunaException ex) {
+            throw ex;
+        } catch (Exception ex) {
+            return EMPTY_PROPERTIES;
```

Comment:
This seems a little blur for the user. What are we trying to recover from? Can we make recoverable cases more explicit?